File: src/Chapter04GraphBFS/CycleDetection.java
package Chapter04GraphBFS;

import java.util.LinkedList;
import java.util.Queue;

public class CycleDetection {
    private Graph G;
    private boolean[] visited;
    private int[] pre;
    private boolean hasCycle = false;

    public CycleDetection(Graph G) {
        this.G = G;
        visited = new boolean[G.V()];
        pre = new int[G.V()];
        for (int i = 0; i < G.V(); i++) {
            pre[i] = -1;
        }

        for (int i = 0; i < G.V(); i++) {
            if (!visited[i]) {
                if (bfs(i)) {
                    hasCycle = true;
                    break;
                }
            }
        }
    }

    public boolean bfs(int s) {
        Queue<Integer> queue = new LinkedList<>();
        queue.add(s);
        visited[s] = true;
        while (!queue.isEmpty()) {
            int v = queue.remove();
            for (int w : G.adj(v)) {
                if (!visited[w]) {
                    queue.add(w);
                    visited[w] = true;
                    pre[w] = v;
                } else {
                    if (w != pre[v]) {
                        return true;
                    }
                }
            }
        }
        return false;
    }

    public boolean hasCycle() {
        return hasCycle;
    }

    public static void main(String[] args) {

        Graph g = new Graph("/Users/hui/Desktop/java/play-with-graph-algorithme/src/Chapter04GraphBFS/g.txt");
        CycleDetection cycleDetection = new CycleDetection(g);
        System.out.println(cycleDetection.hasCycle());

        Graph g2 = new Graph("/Users/hui/Desktop/java/play-with-graph-algorithme/src/Chapter04GraphBFS/g2.txt");
        CycleDetection cycleDetection2 = new CycleDetection(g2);
        System.out.println(cycleDetection2.hasCycle());
    }
}

File: src/Chapter04GraphBFS/python/CycleDetection.py
from Chapter02GraphDFS.python.adjSet import adjSet as Graph
from collections import deque


class CycleDetection:
    def __init__(self, G):
        self._G = G
        self._visited = [False] * G.V
        self._pre = [-1] * G.V
        self.cycle = False
        for v in range(G.V):
            if not self._visited[v]:
                if self._bfs(v):
                    self.cycle = True

    def _bfs(self, s):
        queue = deque()
        queue.append(s)
        self._visited[s] = True
        self._pre[s] = s
        while queue:
            v = queue.popleft()
            for w in self._G.adj(v):
                if not self._visited[w]:
                    queue.append(w)
                    self._visited[w] = True
                    self._pre[w] = v
                elif w != self._pre[v]:
                    self.cycle = True
                    return True
        return False

    @property
    def cycled(self):
        return self.cycle


if __name__ == "__main__":
    filename = "/Chapter04GraphBFS/g.txt"
    g = Graph(filename)

    path = CycleDetection(g)
    print(path.cycled)

    filename = "/Chapter04GraphBFS/g2.txt"
    g = Graph(filename)

    path = CycleDetection(g)
    print(path.cycled)
